docs/rfcs/0007-flow-modules.md

```diff
@@ -0,0 +1,295 @@
+
+## Example
+
+```river
+// module
+argument "password" {
+    optional = false
+    comment = "password for mysql"
+}
+
+argument "username" {
+    optional = false
+    comment = "username for mysql"
+}
+
+export "targets" {
+    comment = "targets for the integration"
+    value = integrations.mysql.server1.targets
+}
+
+integrations.mysql "server1" {
+    username = argument.username.value
+    password = argument.password.value
+}
+
+```
+
+```river
+// parent
+
+local.file "mysql" {
+    filename = "/test/mysql.river"
+}
+
+module.single "mysql" {
+    content = local.file.mysql.content
+    arguments = {
+        {
+            "password" = PASSWORD,
+            "username" = USERNAME,
+        }
+    }
+}
+
+prometheus.scrape "scraper" {
+    targets = module.single.mysql.exports.targets
+}
+
+```
+
+## Limitations
+
+* Duplicate modules cannot be nested, this may or may not be enforced by the system
+* Singleton components are not supported at this time. Example [node_exporter](https://grafana.com/docs/agent/latest/flow/reference/components/prometheus.integration.node_exporter/).
+* Modules will not prevent competing resources, such as starting a server on the same port
+* [Configuration blocks](https://grafana.com/docs/agent/latest/flow/reference/config-blocks/#configuration-blocks) will not be supported. 
+* Arguments and exports within a module must be unique
+
+## Proposal
+
+Add the ability to load `modules` as subgraphs to the primary `graph`. Modules may call other modules within a reasonable stack size depth. Modules are represented as a river string that is interpreted with a defined set of arguments and exports.
+
+The initial component will be `module.single` that will load a single module. Internally these modules will be namespaced so they cannot affect children or parent graphs except via arguments and exports.
+
+Modules will have access to any standard function and any other component exempting singletons. Internally each component in the module will have an `id` that is prepended with the parent's `id` for identification purposes outside of the module. Within the module a component can reference another sibling component normally. There are no known limits on the datatype that a module can use as an argument or export.
+
+`modules.multiple` will use the key of the `map(string)` to uniquely identify a specific module.
+
+### Component Options
+
+
+Given the above example, the `id` of `integrations.mysql "server1"` would be `module.single.mysql.integrations.mysql.server1`. The `data-agent` field would also be prefixed. There are some inherent issues, deeply nested metrics are likely to run into Prometheus label value limits. On Windows platforms there could be issues with the `data-agent` length. These are issues that currently exist in Agent Flow but are more easily hit using deeply nested modules.
+
+
+### Failure Modes
+
+#### Option 1: When a module fails then fail itself and children.
```

Comment:
I do like this approach since it matches up with the idea that modules essentially append sub-graphs onto the overall graph that makes up the Agent configuration, so if the parent breaks you'd expect the failure to cascade down the children. I do see the issue of siblings to the module graph potentially breaking from this, but I guess the call on this is how Flow wants to shape how people should think about the tree of components. 

Option 2 is more relevant to a production use case, though, so in that context Option 2 seems the most correct option for where Flow wants to exist :)